### PR TITLE
add ntcp2proxy support

### PR DESCRIPTION
**Pull request by unlnown542a**
adding ntcp2proxy support

### CODE REVIEW

File: libi2pd/NTCP2.cpp [hunk 1]
```diff
@@ -371,10 +373,82 @@ namespace transport
...
+			memcpy(buff+4, addrbytes.data(), addrsize);
+		}
+		else if (addrtype == eHostname)
+		{
+			buff[3] = 0x03;
+			size_t addrsize = host.size();
+			sz = addrsize + 1 + 4;
+			if (2 + sz > sizeof(buff))
+			{
+				// too big
```
Comment (majestrate) [inline libi2pd/NTCP2.cpp]: maybe you should put a log message here about the host name being too big?

File: libi2pd/NTCP2.cpp [hunk 2]
```diff
@@ -1094,14 +1168,158 @@ namespace transport
...
+		}
+		else
+			LogPrint(eLogError, "NTCP2: unknown proxy type, invalid state");
+	}
+
+	void NTCP2Server::ConnectWithProxy (const std::string& host, uint16_t port, RemoteAddressType addrtype, std::shared_ptr<NTCP2Session> conn)
+	{
+		if(m_ProxyEndpoint == nullptr)
+		{
+			return;
```
Comment (majestrate) [inline libi2pd/NTCP2.cpp]: perhaps make this an assert?

File: libi2pd/NTCP2.h [hunk 3]
```diff
@@ -248,9 +271,15 @@ namespace transport
...
 			std::unique_ptr<boost::asio::ip::tcp::acceptor> m_NTCP2Acceptor, m_NTCP2V6Acceptor;
-			std::map<i2p::data::IdentHash, std::shared_ptr<NTCP2Session> > m_NTCP2Sessions; 
+			std::map<i2p::data::IdentHash, std::shared_ptr<NTCP2Session> > m_NTCP2Sessions;
 			std::list<std::shared_ptr<NTCP2Session> > m_PendingIncomingSessions;
 
+			ProxyType m_ProxyType;
+			std::string m_ProxyAddress;
+			uint16_t m_ProxyPort;
+			boost::asio::ip::tcp::resolver m_Resolver;
+			boost::asio::ip::tcp::endpoint * m_ProxyEndpoint;
```
Comment (majestrate) [inline libi2pd/NTCP2.h]: if it's not owned by anyone else using `std::unique_ptr<boost::asio::ip::tcp::endpoint>` would be easier here. (maybe).

Comment (orignal) [reply libi2pd/NTCP2.h]: I will take care. At least his code works

**Pull request merged**